src/packets.rs

```diff
@@ -232,82 +348,221 @@ impl PacketBuffer {
-        self.data.reserve(additional_size);
-
-        let mut pkt = unsafe {
-            let total_len = self.data.len();
-            self.data.set_len(total_len + additional_size);
-            &mut *(&self.data[total_len] as *const _ as *mut MIDIPacket)
-        };
-
-        pkt.timeStamp = timestamp as MIDITimeStamp;
-        pkt.length = data_len as UInt16;
-        pkt.data[0..data_len].clone_from_slice(&data);
+    pub fn push_packet(&mut self, time: MIDITimeStamp, data: &[u8]) -> &mut Self {
+        let can_merge;
+        let previous_data_len;
+        {
+            // new scope to please the borrow checker
+            let previous_packet = self.last_written_packet();
+            previous_data_len = previous_packet.data().len();
+            can_merge = 
+                previous_packet.timestamp() == time && // timestamps match
+                data[0] != 0xF0 && // not a sysex
+                data[0] & 0b10000000 != 0 && // but first byte is a status byte
+                previous_packet.data()[0] != 0xF0 && // previous packet not a sysex
+                previous_packet.data()[0] & 0b10000000 != 0 && // but first byte is a status byte
+                previous_data_len + data.len() < MAX_PACKET_DATA_LENGTH; // enough space left in the packet
+        }
 
-        let mut pkt_list = unsafe { &mut *(self.data.as_mut_ptr() as *mut MIDIPacketList) };
-        pkt_list.numPackets += 1;
-        self.packet_list = PacketList(pkt_list);
+        if can_merge {
+            // write the data into the previous packet
+            unsafe {
+                self.storage.set_len(self.last_written_pkt_offset + PACKET_HEADER_SIZE + previous_data_len + data.len());
+                self.storage.extend_packet(self.last_written_pkt_offset, data);
+            }
+        } else  {
+            let next_offset = self.get_next_offset();
+            unsafe {
+                self.storage.set_len(next_offset + PACKET_HEADER_SIZE + data.len());
+                self.storage.assign_packet(next_offset, time, data);
+                self.storage.deref_mut().inner.num_packets += 1;
+            }
+            self.last_written_pkt_offset = next_offset;
+        }
 
         self
     }
-}
 
-impl Deref for PacketBuffer {
-    type Target = PacketList;
-
-    fn deref(&self) -> &PacketList {
-        &self.packet_list
+    #[inline]
+    fn last_written_packet(&self) -> &Packet {
+        // NOTE: This requires that there always is at least one packet in the buffer
+        //       (which is okay because we do not provide an empty constructor)
+        unsafe { &*(self.storage.get_slice()[self.last_written_pkt_offset..].as_ptr() as *const Packet) }
+    }
+    
+    #[inline]
+    fn get_next_offset(&self) -> usize {
+        let length = self.last_written_packet().inner.length as usize;
+        let next_unadjusted = self.last_written_pkt_offset + PACKET_HEADER_SIZE + length;
+        if cfg!(any(target_arch = "arm", target_arch = "aarch64")) {
```

Comment:
I've seen this check several times by now. If by any chance we need to cover a new architecture that requires u32 alignment, we will need to change several of those checks which is prone to error. Do you think it could be extracted into a macro and used all around ?

```rust
if align_u32!() { ... }
```

Reply:
I basically copied the logic from [here](https://github.com/phracker/MacOSX-SDKs/blob/9fc3ed0ad0345950ac25c28695b0427846eea966/MacOSX10.13.sdk/System/Library/Frameworks/CoreMIDI.framework/Versions/A/Headers/MIDIServices.h#L2098-L2103). Trying to think about a way to extract this that works at the expression level (i.e. `if cfg!(...) { ... }`) and also on the item definition level (i.e. `#[cfg(...)]`), I can only come up with this:
```
#[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
mod alignment {
    type AlignmentMarker = [u32; 0];
    const NEEDS_ALIGNMENT: bool = true;
}

#[cfg(not(any(target_arch = "arm", target_arch = "aarch64")))]
mod alignment {
    type AlignmentMarker = [u8; 0];
    const NEEDS_ALIGNMENT: bool = false;
}```

Would that be okay?

Reply:
That's a very interesting approach. I like it.